### PR TITLE
Complete MSStore download and licensing

### PR DESCRIPTION
**Pull request by yao-msft**
The MSStore download feature is mainly targeted for enterprise offline distribution. So a Microsoft Entra Id account that is at least Administrator role of the directory is required for licensing. A follow up pr will make the DisplayCatalog calls (for acquiring MSStore package info) require Microsoft Entra Id account as well, when the backend configuration is done.

Added tests. Manually validated and applied feedback with stakeholders too.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4464)

### CODE REVIEW

**Review by ryfu-msft — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/Public/winget/MSStoreDownload.h [hunk 1]
```diff
@@ -12,76 +14,43 @@
...
-        // Gets the best installer from the manifest, if at least one is applicable.
-        std::optional<MSStoreCatalogPackage> GetPreferredPackage(const std::vector<MSStoreCatalogPackage>& package);
-
-        // Determines if an installer is applicable.
-        bool IsApplicable(const MSStoreCatalogPackage& package);
-
-        //// Determines if the first installer is a better choice.
-        bool IsFirstBetter(const MSStoreCatalogPackage& first, const MSStoreCatalogPackage& second);
+        // Gets license for the corresponding package returned by previous GetDownloadInfo().
+        // GetDownloadInfo() must be called before calling this method.
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/winget/MSStoreDownload.h:45]: This comment suggests that this method should be on the download info object instead...

File: src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp [hunk 2]
```diff
@@ -64,6 +73,70 @@ namespace AppInstaller::CLI::Workflow
 
             THROW_IF_FAILED(hr);
         }
+
+        HRESULT DownloadMSStorePackageFile(const MSStore::MSStoreDownloadFile& downloadFile, const std::filesystem::path& downloadDirectory, Execution::Context& context)
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp:77]: It feels like there should be some progress and cancellation handling.

File: src/AppInstallerCommonCore/MSStoreDownload.cpp [hunk 3]
```diff
@@ -4,488 +4,1045 @@
...
+        // {
+        //   "license": {
+        //     "keys": [ // returned as array for future, for now only 1 key
+        //       {
+        //         "value": "base64 encoded string"
+        //       }
+        //     ]
+        //   }
+        // }
+        std::vector<BYTE> GetLicencing(std::string_view contentId, const Http::HttpClientHelper::HttpRequestHeaders& authHeaders)
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/MSStoreDownload.cpp:930]: Typo?
```suggestion
        std::vector<BYTE> GetLicensing(std::string_view contentId, const Http::HttpClientHelper::HttpRequestHeaders& authHeaders)
```

File: src/AppInstallerCommonCore/MSStoreDownload.cpp [hunk 4]
```diff
@@ -4,488 +4,1045 @@
...
+#endif
+            Authentication::MicrosoftEntraIdAuthenticationInfo licensingMicrosoftEntraIdAuthInfo;
+            licensingMicrosoftEntraIdAuthInfo.Resource = "c5e1cb0d-5d24-4b1a-b291-ec684152b2ba";
+            Authentication::AuthenticationInfo licensingAuthInfo;
+            licensingAuthInfo.Type = Authentication::AuthenticationType::MicrosoftEntraId;
+            licensingAuthInfo.MicrosoftEntraIdInfo = std::move(licensingMicrosoftEntraIdAuthInfo);
+
+            // Not moving authArgs because we'll have auth for display catalog and sfs client in the near future.
+            m_licensingAuthenticator = std::make_unique<Authentication::Authenticator>(std::move(licensingAuthInfo), authArgs);
+#ifndef AICLI_DISABLE_TEST_HOOKS
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/MSStoreDownload.cpp:1025]: style: You can leave the `{ }` in and just #def out the `if` line.  That way you don't need a second #ifdef.

File: src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp (hunk 2 above)
Comment (yao-msft) [reply src/AppInstallerCLICore/Workflows/MSStoreInstallerHandler.cpp:77]: DownloadPackageInstaller workflow task (within this method) handles progress, cancellation and retry for each download. I think we could improve in the future if this experience is not enough.

**Review by ryfu-msft — APPROVED** [review]

**Pull request merged**